### ISSUE TITLE
[FEA] Automated canonicalization tests

### ISSUE THREAD

**Issue by jlowe** [NVIDIA/spark-rapids#1309]
**Is your feature request related to a problem? Please describe.**
One of the causes of #1308 is parts of the plan failing to canonicalize properly, and this confuses the Spark query planner into thinking it cannot reuse parts of the query that are reusable.  This ends up running scans redundantly during the query needlessly.

The canonicalization error was subtle and easy to miss, so we need a better way of automatically detecting when canonicalization methods are incorrect.

**Describe the solution you'd like**
One idea is to inject canonicalization testing into our existing unit and integration tests.  When running a GPU query, we could plan the query _twice_ and verify that the two executed plans canonicalized are equivalent.

For example, if we have a plan like `spark.sql(query).collect` we could do something like this:
```scala
val plan1 = spark.sql(query)
val plan2 = spark.sql(query)
val isCanonical = plan1.queryExecution.executedPlan.canonicalized == plan2.queryExecution.executedPlan.canonicalized
plan1.collect
```

**Describe alternatives you've considered**
If we could programmatically generate unit tests for the various combinations of GPU Exec nodes and expressions that would be great, but I think artificially creating some of the input arguments outside of a query could be tricky.

**Comment by sameerz** [comment]
@andygrove does this issue get resolved by PR #1349 or do we need to wait for https://issues.apache.org/jira/browse/SPARK-33744 to close this?

**Comment by andygrove** [comment]
Yes, this was resolved by #1349. We skipped the canonicalization test on 4 unit tests due to SPARK-33744.
